docs/shared/concepts/customizing-inputs.md

```diff
@@ -0,0 +1,160 @@
+}
+```
+
+This configuration defines three `namedInputs`.
+
+- `sharedGlobals` contains root level files that should invalidate the cache for all projects. Note that `tsconfig.base.json` and the `package.json` lock files are handled behind the scenes by Nx so that when you modify the specific properties that affect your project, the cache is invalidated.
+- `default` contains every file in the project and everything defined in `sharedGlobals`
+- `production` starts with every file in `default`, but removes `.spec.ts` files, markdown files and some project-specific config files.
+
+These names are not hard-coded. You can add your own `namedInputs` or rename these as you see fit.
+
+With these `namedInputs` in place, we can set default `inputs` for project targets in `nx.json`:
+
+```jsonc
+{
+  "namedInputs": {
+    "sharedGlobals": [
+      /* ... */
+    ],
+    "default": [
+      /* ... */
+    ],
+    "production": [
+      /* ... */
+    ]
+  },
+  "targetDefaults": {
+    "build": {
+      "inputs": ["production", "^production"],
+      "dependsOn": ["^build"]
+    },
+    "test": {
+      "inputs": ["default", "^production", "{workspaceRoot}/jest.preset.js"]
+    }
+  }
+}
+```
+
+In this configuration, `build` needs to be re-run if the project's `production` files are changed or the dependencies' `production` files are changed. The `"dependsOn": ["^build"]` setting ensures that the output files of the dependencies are correct, but it does not tell Nx to re-run this project's build target when those outputs change. That's why `"^production"` is needed in the `inputs` array.
+
+Note that we are using `^production` instead of `^default` because changing test files of a dependency should not invalidate the build cache for this project.
+
+The `test` target cache is invalidated if test or production code is changed in this project (`default`) or if production code is changed in a dependency `^production`. The tests also need to be run again if the root level `jest.preset.js` file is modified.
+
+### Project Settings
+
+In the `ui-library` library, we have some `*.spec.tsx` files that we don't want to count as production files. We can modify the project configuration to account for this:
+
+```jsonc
+{
+  "namedInputs": {
+    "production": [
+      // copied this definition from nx.json
+      "default",
+      "!{projectRoot}/**/*.spec.ts",
+      "!{projectRoot}/**/*.spec.tsx", // added this line
+      "!{projectRoot}/**/*.md",
+      "!{projectRoot}/tsconfig.spec.json",
+      "!{projectRoot}/jest.config.ts",
+      "!{projectRoot}/.eslintrc.json"
+    ]
+  }
+}
+```
+
+Note that I need to copy the entire definition of `production` from `nx.json`. This entry overwrites the existing `production` entry defined in `nx.json`, but `default` and `sharedGlobals` are inherited. Now when I modify a `spec.tsx` file in `ui-library`, the build for `ui-library` is unaffected and the build for `react-app` is unaffected.
+
+{% callout type="note" %}
```

Comment:
Can you please add a title for callouts?

Reply:
done